### DISCUSSION TITLE
Run multiple scripts for a single package

### DISCUSSION THREAD

**Discussion by kettanaito** [vercel/turborepo#636] [Help]
Hi! Thanks for creating a better way of managing monorepos!

I've got a repository that needs 2 scripts to run in parallel in order to work on it: `dev` and `dev:preview`. How can I configure Turborepo so when I run the root-level `npm run dev` (`turbo run dev --parallel`), it automatically runs both scripts in one of the repos?

## Things I've tried

I've tried adding a new `dev:preview` command to the pipeline but that had no effect. 

```json
{
  "baseBranch": "origin/main",
  "pipeline": {
    "build": {
      "dependsOn": ["^build"],
      "outputs": ["dist/**", ".next/**"]
    },
    "lint": {
      "outputs": []
    },
    "dev": {
      "cache": false
    },
    "dev:preview": {
      "cache": false,
      "dependsOn": ["dev"]
    }
  }
}
```

```sh
turbo run dev dev:preview --parallel
```

**Comment by leonardssh** [comment]
I've tested what you explained above, and it works perfectly. 

I've left below an example of how to configure `package.json` files and `turbo.json`.
```jsonc
// turbo.json
{
  "baseBranch": "origin/main",
  "pipeline": {
    "build": {
      "dependsOn": ["^build"],
      "outputs": ["dist/**", ".next/**"]
    },
    "lint": {
      "outputs": []
    },
    "dev": {
      "cache": false
    },
    "dev:preview": {
      "cache": false,
      "dependsOn": ["dev"]
    }
  }
}
```

```jsonc
// root package.json
{
	"name": "root",
	"private": true,
	"workspaces": [
		"packages/*"
	],
	"scripts": {
		"build": "turbo run build",
		"dev": "turbo run dev",
                "dev:preview": "turbo run dev:preview",
		"lint": "turbo run lint",
	},
	"packageManager": "pnpm@6.26.1"
}
```
```jsonc
// target package.json from packages
{
	"name": "MY_PACKAGE_NAME",
	"scripts": {
		"build": "YOUR BUILD SCRIPT",
		"dev": "YOUR DEV SCRIPT",
                "dev:preview": "YOUR DEV:PREVIEW SCRIPT",
		"lint": "YOUR LINT SCRIPT",
	}
}
```

If it still doesn't work, give me some reproduction code to see your environment.

**Reply by kettanaito** [reply]
Thanks for a quick reply! 
Based on the code above, I think you need to run two commands in separate processes:

- `npm run dev`
- `npm run dev:preview`

Is that correct? 

I'd like to have a single root-level command `npm run dev` and tell turborepo to run multiple commands for some repositories. That's why I tried `turbo run dev dev:preview --parallel`, which didn't work for some reason. 

Perhaps worth mentioning that not all repos will have `dev:preview`.

**Reply by leonardssh** [reply]
It's not necessary that all repos have `dev:preview`, I've made a small example here to give you an idea. https://github.com/LeonardSSH/turborepo-discussions-636

Once you have specified that `dev:preview` dependsOn `dev`, running `pnpm run dev:preview` will run the `dev` script first and then `dev:preview`.

Reference: https://turborepo.org/docs/features/pipelines#defining-a-pipeline

**Reply by leonardssh** [reply]
Oh, now I realize exactly what the problem is, let me get this straight, you want to run multiple scripts for only one package?

To do this, you can use `--scope`.

e.g 
```sh
> turbo run dev dev:preview --scope="package-name"
```

**Reply by leonardssh** [reply]
Also in the future we'll benefit from `--filter` (smth like pnpm filter stuff). There's a PR related to it https://github.com/vercel/turborepo/pull/81.

With the help of which we can do something like:

```json
"scripts": {
    "server": "turbo run --filter=\"server\"",
    "client": "turbo run --filter=\"client\""
}
```
Usage:
```sh
> pnpm run server <action>
> pnpm run client build
```

**Reply by kettanaito** [reply]
Thanks for preparing an example 😮 I will take a look!

No, my idea is not to run it in a single package but _for_ a single package. 

```
packages
  - a
    - dev
  - b
    - dev
    - dev:preview
  - c
    - dev
```

When I run `npm run dev` I expect for the following scripts to be run in parallel:

- `a:dev:`
- `b:dev:`
- `b:dev:preview:`
- `c:dev:`

**Reply by leonardssh** [reply]
I understand what you mean now, I don't think it is possible at the moment to run such tasks. I mean with pipelines you can only do dependsOn tasks. (e.g not run script x until script y has run)

You can do some hardcoding and put a `"postdev": "..."` and it will run after the `dev` has run.

**Reply by leonardssh** [reply]
Take a look at https://github.com/vercel/turborepo/issues/591, it seems to use some more interesting dependsOn, maybe it's what you're looking for.

**Reply by fbartho** [reply]
I know I'm necromancing this thread, but I found a workaround: 
1. create a peer package as a stub, basically all it needs is a `name`, a `scripts`, and a boilerplate `turbo.json`
2. make the original package have this new package as a dependency (this is for the build-script to auto-execute)
3. make the build script on this new package call into the original package's script alias for the second operation, this ensures the peer-package will wait for the generated file before its first build in watch-mode. (important if you `.gitignore` the generated files)
4. make the dev script on this new package call into the original package's script alias for the second operation (the watch alias)

```json
{
	"name": "@repo/secondary-watcher-stub",
	"description": "Stub-package to watch graphql files for ../peer-package",
	"scripts": {
		"build": "cd ../peer-package && yarn build:graphql",
		"dev": "cd ../peer-package && yarn dev:graphql"
	}
}
```

**Reply by selvinkuik** [reply]
@fbartho Impecably necromancing timing. 2025 and this works like a charm!
